community/cypher/cypher-compiler-3.2/src/main/scala/org/neo4j/cypher/internal/compiler/v3_2/planner/logical/plans/rewriter/distinctVarExpander.scala

```diff
@@ -0,0 +1,88 @@
+/*
+ * Copyright (c) 2002-2016 "Neo Technology,"
+ * Network Engine for Objects in Lund AB [http://neotechnology.com]
+ *
+ * This file is part of Neo4j.
+ *
+ * Neo4j is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+package org.neo4j.cypher.internal.compiler.v3_2.planner.logical.plans.rewriter
+
+import org.neo4j.cypher.internal.compiler.v3_2.planner.logical.plans._
+import org.neo4j.cypher.internal.frontend.v3_2.ast.{Expression, FunctionInvocation}
+import org.neo4j.cypher.internal.frontend.v3_2.{Rewriter, topDown}
+
+import scala.collection.mutable
+
+case object distinctVarExpander extends Rewriter {
+
+  private def findDistinctSet(plan: LogicalPlan): Set[LogicalPlan] = {
+
+    def collectDistinctSet(plan: LogicalPlan,
+                           distinctBy: Option[Set[String]],
+                           distinctSet: mutable.Set[LogicalPlan]): Unit = {
+      val lowerDistinctLand: Option[Set[String]] = plan match {
+        case Aggregation(left, groupExpr, aggrExpr) if aggrExpr.values.forall(isDistinct) =>
+          val variablesInTheDistinctSet = groupExpr.values.flatMap(_.dependencies.map(_.name)).toSet
+          Some(variablesInTheDistinctSet)
+
+        case expand: VarExpand
+          if distinctBy.nonEmpty && !distinctNeedsRelsFromExpand(distinctBy, expand) =>
+          distinctSet += plan
```

Comment:
In this version of collect we are only putting `LogicalPlan`s that are `VarExpand`. Could return a more specific type if we want. On the other hand, I doing the general implementation where all distinct LogicalPlans are put into the returned set might encourage future disctinct specializations.